src/serial_io.rs

```diff
@@ -1,41 +1,110 @@
-use esp32c3_hal::{
-    interrupt,
-    interrupt::CpuInterrupt,
-    pac::{self, UART0},
-    prelude::*,
-    serial::Instance,
-    Cpu, Serial,
+use heapless::Deque;
+#[cfg(any(target_arch = "riscv32"))]
+use riscv::interrupt::free as interrupt_free;
+#[cfg(any(target_arch = "xtensa"))]
+use xtensa_lx::interrupt::free as interrupt_free;
+
+use crate::{
+    hal::{
+        interrupt,
+        interrupt::CpuInterrupt::*,
+        pac,
+        pac::UART0,
+        prelude::*,
+        serial::Instance,
+        Cpu::*,
+        Serial,
+    },
+    protocol::InputIO,
 };
-use heapless::spsc::Queue;
 
-use crate::{protocol::InputIO, targets::esp32c3 as target};
+const RX_QUEUE_SIZE: usize = crate::targets::MAX_WRITE_BLOCK + 0x400;
 
-const RX_QUEUE_SIZE: usize = target::MAX_WRITE_BLOCK + 0x400;
-static mut RX_QUEUE: Queue<u8, RX_QUEUE_SIZE> = Queue::new();
+static mut RX_QUEUE: Deque<u8, RX_QUEUE_SIZE> = Deque::new();
 
-impl<T: Instance> InputIO for Serial<T> {
+impl<'a, T: Instance> InputIO for Serial<T> {
     fn recv(&mut self) -> u8 {
-        loop {
-            if let Some(byte) = unsafe { RX_QUEUE.dequeue() } {
-                return byte;
-            }
-        }
+        unsafe { while interrupt_free(|_| RX_QUEUE.is_empty()) {} }
+        unsafe { interrupt_free(|_| RX_QUEUE.pop_front().unwrap()) }
     }
 
     fn send(&mut self, bytes: &[u8]) {
         self.write_bytes(bytes).unwrap()
     }
 }
 
-#[interrupt]
-fn UART0() {
+fn uart_isr() {
     let uart = unsafe { &*UART0::ptr() };
 
     while uart.status.read().rxfifo_cnt().bits() > 0 {
         let data = uart.fifo.read().rxfifo_rd_byte().bits();
-        unsafe { RX_QUEUE.enqueue(data).unwrap() };
+        unsafe { RX_QUEUE.push_back(data).unwrap() };
     }
 
     uart.int_clr.write(|w| w.rxfifo_full_int_clr().set_bit());
-    interrupt::clear(Cpu::ProCpu, CpuInterrupt::Interrupt3);
+}
+
+#[cfg(feature = "esp32")]
+pub fn enable_uart0_rx_interrupt() {
```

Comment:
We have the ability to enable serial interrupts in esp-hal now, so we can remove these functions now. Though we may still need the interrupt mapping parts.

Reply:
It's referring to https://github.com/esp-rs/esp-hal/issues/197